### PR TITLE
docs(api): update custom hooks API for plugins

### PR DESCRIPTION
**Pull request by snitin315**
Resolve https://github.com/webpack/webpack.js.org/issues/5583

### CODE REVIEW

File: src/content/api/plugins.mdx [hunk 1]
```diff
@@ -87,18 +88,50 @@ The moral of the story is that there are a variety of ways to `hook` into the
...
 ## Custom Hooks
 
-In order to add a new hook to the compilation for other plugins to `tap` into,
-you then `require` the necessary hook class from `tapable` and create one:
+In order to offer a custom hook to the compilation for other plugins to `tap` into,
+you need to do the following:
 
-```js
-const SyncHook = require('tapable').SyncHook;
+- Create a module-scope `WeakMap` for compilation hooks:
```
Comment (chenxsan) [inline src/content/api/plugins.mdx:94]: Let's use ordered list here.

File: src/content/api/plugins.mdx [hunk 2]
```diff
@@ -87,18 +88,50 @@ The moral of the story is that there are a variety of ways to `hook` into the
...
 
-if (compiler.hooks.myCustomHook) throw new Error('Already in use');
-compiler.hooks.myCustomHook = new SyncHook(['a', 'b', 'c']);
+  ```ts
+  const compilationHooks = new WeakMap<Compilation, MyHooks>();
```
Comment (chenxsan) [inline src/content/api/plugins.mdx:97]: Will it make more sense to use js instead of ts in this page?

Comment (snitin315) [reply src/content/api/plugins.mdx:97]: I feel it's ok to use typescript here as the usage is basic.

File: src/content/api/plugins.mdx [hunk 3]
```diff
@@ -87,17 +88,51 @@ The moral of the story is that there are a variety of ways to `hook` into the
...
+static getCompilationHooks(compilation: Compilation) : MyHooks {
+  let hooks = compilationHooks.get(compilation);
+  if(hooks === undefined) {
+    compilationHooks.set(compilation, hooks = {
+      custom: new SyncHook()
+    });
+  }
+  return hooks;
+}
+```
```
Comment (chenxsan) [inline src/content/api/plugins.mdx:116]:
```suggestion
    ```ts
    static getCompilationHooks(compilation: Compilation) : MyHooks {
      let hooks = compilationHooks.get(compilation);
      if(hooks === undefined) {
        compilationHooks.set(compilation, hooks = {
          custom: new SyncHook()
        });
      }
      return hooks;
    }
    ```
```

File: src/content/api/plugins.mdx [hunk 4]
```diff
@@ -87,17 +88,51 @@ The moral of the story is that there are a variety of ways to `hook` into the
...
-compiler.hooks.myCustomHook = new SyncHook(['a', 'b', 'c']);
+3. Call hooks like below in your plugin:
 
-// Wherever/whenever you'd like to trigger the hook...
-compiler.hooks.myCustomHook.call(a, b, c);
+```ts
+const hooks = MyPlugin.getCompilationHooks(compilation);
+
+hooks.custom.call(1, 'hello');
+```
```
Comment (chenxsan) [inline src/content/api/plugins.mdx:124]:
```suggestion
    ```ts
    const hooks = MyPlugin.getCompilationHooks(compilation);

    hooks.custom.call(1, 'hello');
    ```
```

File: src/content/api/plugins.mdx [hunk 5]
```diff
@@ -87,17 +88,51 @@ The moral of the story is that there are a variety of ways to `hook` into the
...
+
+```ts
+import MyPlugin from 'my-plugin';
+
+const hooks = MyPlugin.getCompilationHooks(compilation);
+
+hooks.custom.tap('OtherPlugin', (n, s) => {
+  // magic
+});
 ```
```
Comment (chenxsan) [inline src/content/api/plugins.mdx:136]:
```suggestion
    ```ts
    import MyPlugin from 'my-plugin';

    const hooks = MyPlugin.getCompilationHooks(compilation);

    hooks.custom.tap('OtherPlugin', (n, s) => {
      // magic
    });
    ```
```

**Review by chenxsan — APPROVED** [review]

**Pull request merged**